libnavui-app/src/main/java/com/mapbox/navigation/ui/app/internal/controller/RoutePreviewStateController.kt

```diff
@@ -109,48 +121,63 @@ class RoutePreviewStateController(private val store: Store) : StateController()
             }
     }
 
-    private suspend fun MapboxNavigation.fetchRoute(
-        routeOptions: RouteOptions,
-        fetchStarted: (requestId: Long) -> Unit
-    ): List<NavigationRoute> {
-        return suspendCancellableCoroutine { cont ->
-            val requestId = requestRoutes(
-                routeOptions,
-                object : NavigationRouterCallback {
-                    override fun onRoutesReady(
-                        routes: List<NavigationRoute>,
-                        routerOrigin: RouterOrigin
-                    ) {
-                        cont.resume(routes)
-                    }
-
-                    override fun onFailure(
-                        reasons: List<RouterFailure>,
-                        routeOptions: RouteOptions
-                    ) {
-                        cont.resumeWithException(FetchRouteError(reasons, routeOptions))
+    // This SAGA dispatches FetchRoute action, awaits RoutePreviewState.Ready
+    // and updates NavigationState to either RoutePreview or ActiveNavigation state.
+    private suspend fun fetchRouteAndMoveToNavigationStateSaga() {
+        store.actionsFlowable()
+            .filter {
+                it is RoutePreviewAction.FetchRouteAndShowRoutePreview ||
+                    it is RoutePreviewAction.FetchRouteAndStartActiveNavigation ||
+                    it is NavigationStateAction.Update
+            }
+            .collectLatest {
+                when (it) {
+                    is RoutePreviewAction.FetchRouteAndShowRoutePreview -> {
+                        if (fetchRouteIfNeeded()) {
+                            store.dispatch(
+                                NavigationStateAction.Update(NavigationState.RoutePreview)
+                            )
```

Comment:
I like how `ThunkActions` and `BackPressedComponent` are the only entities that dispatch `NavigationStateAction.Update(*)`. Similar to how you are doing with `is RoutePreviewAction.FetchRouteAndStartActiveNavigation` below where you dispatch `startActiveNavigation(previewRoutes.routes)` can we do the same here and dispatch 
```
showRoutePreview(
    store.state.value.destination?.point!!,
    (store.state.value.previewRoutes as RoutePreviewState.Ready).routes
)
```
I understand that this will set the same `destination` and `routes` to the state which already holds it, but in my opinion it's not that bad to set the same value given that now we only have single entities firing `Update` action.
Wdyt?

Reply:
We could, but force unwraps and unsafe casts concern me.

Right now, we manually move `NavigationState` by dispatching `NavigationStateAction.Update(...)` after we set the Store State values. Unfortunately, this error-prone approach can easily lead to a `NavigationState` desync when we forget to update required values. So I prefer to leave the above as is and focus on refactoring how `NavigationState` is managed. First, we should introduce a state machine (inside `NavigationStateController`) that would observe the Store and MapboxNavigation and derive the `NavigationState` from currently available values.